kmp_cpu.cpp

```diff
@@ -3,30 +3,66 @@
 static void get_fail(const char *pattern, int pattern_length, int *fail) {
     int candidate = 0;
     fail[0] = -1;
+    unsigned char clear_mask = 0x00000011;
     for (int pos = 1; pos < pattern_length; pos++, candidate++) {
-        if (pattern[pos] == pattern[candidate]) {
+        unsigned char com_pos = (pattern[pos / 4] >> (2 * (pos % 4))) | clear_mask;
+        unsigned char com_cand = (pattern[candidate / 4] >> (2 * (candidate % 4))) | clear_mask;
+        if (com_pos == com_cand) {
             fail[pos] = fail[candidate];
         }
         else {
             fail[pos] = candidate;
-            while (candidate >= 0 && pattern[pos] != pattern[candidate]) {
+            while (candidate >= 0 && com_pos != com_cand) {
                 candidate = fail[candidate];
             }
         }
     }
     fail[pattern_length] = candidate;
 }
 
+void char_compress(const char *text, int text_length, char& text_com) {
+    // A: 00; T: 01; C:10; G: 11 
+    unsigned char A_mask = 0x00000000;
+    unsigned char T_mask = 0x00000001;
+    unsigned char C_mask = 0x00000010;
+    unsigned char G_mask = 0x00000011;
```

Comment:
All these masks should also use `0b` prefix instead of `0x`.

Reply:
changed